Refactor Activators common code and improve its testability

tomekl007 opened the pull request:
I've created the `BaseActivator` that encapsulates activator logic.
It solves a use case when the service needs to wait for N other services. 
This pattern is used in the config-store and Kafka-CDC-producer.
I think that many client's modules may follow the same pattern. 
Let me know wdyt.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+    this.context = context;
+
+    for (Class<?> dependentService : dependentServices) {
+      ServiceReference<?> serviceReference =
+          context.getServiceReference(dependentService.getName());
+      if (serviceReference == null) {
+        logger.debug(
+            "{} service is null, registering a listener to get notification when it will be ready.",
+            dependentService.getName());
```
I guess this may be subject to a (low probability) race since we do not know when exactly the dependent service may be registered.

What do you think about using `org.osgi.util.tracker.ServiceTracker` instead?

tomekl007 replied:
Very good idea, done.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+            "{} service is null, registering a listener to get notification when it will be ready.",
+            dependentService.getName());
+        context.addServiceListener(
+            this, String.format("(objectClass=%s)", dependentService.getName()));
```
Could we support richer filters? For example, services that depend on `Persistence` use filter like `(Identifier=%s)` instead of the class name.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+   * @return ServiceAndProperties that has the service for OSGi registration and the properties that
+   *     will be passed.
+   */
+  protected abstract ServiceAndProperties createService(List<Object> dependentServices);
```
Could we provide type safe references to dependencies? One possible way is to take "service selector" objects in the constructor and populate them when the services become available.

tomekl007 replied:
I don't see a way of making it more compile-time type-safe without reducing the extensibility of this base class. 
I think that at some point we need to operate on `Object`.
Could you elaborate more (give example) how would you make it type-safe?

dimas-b replied:
Rough sketch:

```

class BaseActivator {
protected abstract Collection<ServicePointer<?>> dependencies();
... 
}

class ServicePointer<T> {

  ServicePointer(Class<T> targetClass, String filter) {...}

  // called by BaseActivator
  setService(T service) {...}

 // called by user (sub-class of BaseActivator)
T getService() {...}
}
```

tomekl007 replied:
at some point, you will need to invoke `  setService(T service) {...}` that takes T, but the service retrieved here:
`public Object addingService(ServiceReference<Object> ref)`
is of an `Object` type, so it is impossible to call `setService(T)`.

Second issue - the dependencies must be of a `?` type. Because of that, when the sub-class of `BaseActivator` will get it, it is still impossible to have the type-safety. You will still need to cast:
```
@Override
  protected ServiceAndProperties createService(List<ServicePointer<?>> dependentServices) {
	  Object service = dependentServices.get(0).getService(); // need to cast
            ...
```
So I don't see how the ServicePointer will solve this issue.

dimas-b replied:
Yes, I think it's unavoidable in the java type system. However, I think it is preferable to have the base / utility class deal with type casts so that sub-classes (majority) will not have to do type casts. 

I think we can also check the actual type of the service object before injecting it into `ServicePointer` since we that the `Class` objects for its expected type (e.g. use `Class.cast(...)`).

tomekl007 replied:
The check before injecting already exists:
```
  for (Map.Entry<Class<?>, Object> registeredService : registeredServices.entrySet()) {
        if (registeredService.getKey().isAssignableFrom(service.getClass())) {
          logger.debug("{} using service: {}", activatorName, ref.getBundle());
          registeredServices.put(registeredService.getKey(), service);
        }
      }
```

Regarding hiding it in the parent class, of course, it would be better. Unfortunately, even with your approach, sub-classes will need to cast. The reason for this is the wildcard `?`:
```
Collection<ServicePointer<?>> dependencies();
```
as long as you have one service pointer, it will work:
```
ServicePointer<String> stringServicePointer = new ServicePointer<>(String.class);
String service = stringServicePointer.getService(); // works
```
but please keep in mind that it will be a list of different `T` so for example:
```
List<ServicePointer<?>> dependencies = Arrays.asList(
				new ServicePointer<>(String.class),
				new ServicePointer<>(Integer.class)
		);
```
event if every single `ServicePointer` is type-safe the caller will still operate on `Object`:
```
		Object service1 = dependencies.get(0).getService();
		Object service2 = dependencies.get(0).getService();
```
and do the cast. Using `cast` does not solve this issue:
```
Object cast = dependencies.get(0).targetClass.cast(dependencies.get(0).getService());
```
So as long as the client needs to know exact types (and those are all our cases), the cast MUST be in the client code. wdyt?

dimas-b replied:
the `?` is only at the base class level. Sub-classes will not their types, e.g.

```
class ExampleServiceActivator extends BaseActivator {

  private final ServiceReference<Persistence> persistence = new ...

 @Override
 protected Collection<ServiceReference<?>> dependencies() {
   // this may need to be adjusted... not sure about exact signature, but should work in general, I believe
   return Collections.singletonList(persistence)
 }

 @Override
 void startServioce() { // or whatever 
   MyService = new MyService(persistence.get());
 }
}

class MyService {
 MyService(Persistence p) {...}
}
```

tomekl007 replied:
That was a very good idea! Thank you. I've implemented the type-safe variant of this API using your approach, please take a look.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+  @Override
+  public synchronized void serviceChanged(ServiceEvent event) {
+    for (Class<?> dependentService : dependentServices) {
+      Object registeredService = BundleUtils.getRegisteredService(context, event, dependentService);
```
what about dependent service de-registration?

tomekl007 replied:
I've reworked this example to use `Tracker`.
It already provides the way to unget service:
```
	public void removedService(ServiceReference<S> reference, T service) {
		context.ungetService(reference);
	}
```
So I think that we don't need to handle it explicitly.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+    started = true;
+    ServiceAndProperties service = createService(dependentServices);
+    context.registerService(targetServiceClass.getName(), service.service, service.properties);
+    logger.info("Started {}....", activatorName);
```
nit: `Started` and `...` kind of contradict each other :)

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+    }
+    started = true;
+    ServiceAndProperties service = createService(dependentServices);
+    context.registerService(targetServiceClass.getName(), service.service, service.properties);
```
Is there any particular reason to use the class _name_ for service registration as opposed to the class _object_?

dougwettlaufer replied:
Also, what about things that don't need to be registered? For example, a top level service like graphql doesn't need to register itself.

tomekl007 replied:
done in: https://github.com/stargate/stargate/pull/370/commits/47f882cccf8a281b5680d272bf5d8f372b86f0ec

tomekl007 replied:
>
Is there any particular reason to use the class name for service registration as opposed to the class object?

the `registerService` API is not accepting the `targetServiceClass` of a `Class<?> type` as an argument.

dougwettlaufer commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  @Override
+  public synchronized void stop(BundleContext context) throws Exception {
```
In many cases this will need to be overwritten.

tomekl007 replied:
I've added the `protected abstract void stopService();`  that should be called by the client

tomekl007 commented:
@dimas-b @dougwettlaufer I've changed the design to use `ServiceTracker` - it looks better now.
I've addressed all comments and added missing features. I also migrated `GrapqlActivator` to a new mechanism. Let me know wdyt.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,255 @@
+    }
+  }
+
+  public static class DependentService {
```
What do you think about renaming to `ServiceDeprendency` for the sake of clarify? I believe "dependent" refers to the entity that depends on something else, but this class is meant to represent that "something" on which our services depend.

Also, in another comment I suggested a using a class named like `ServicePointer`, enhanced with the ability to pass type-checked references to the dependent service.

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,255 @@
+  public static class DependentService {
+    private Class<?> className;
+
+    private Optional<String> identifier;
```
How about allowing any OSGi filter for generalizability?

tomekl007 replied:
I am not sure what you mean. The filter is used at a different level, see:
`new Tracker(
            context, context.createFilter(constructDependenciesFilter()), registeredServices);`
and it is constructed from a `String` (we are doing this for all other activators).

dimas-b replied:
I mean allowing BaseActivator sub-classes to provide a general-purpose OSGI filter, not just the service identifier.

tomekl007 replied:
I see your point but I don’t see a use case for it. We have only two cases: exact class and identifier. Both are supported by the current solution. If there will be a new use case, we can easily parameterize it or expose filter.

dimas-b replied:
ok

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,255 @@
+      this.identifier = Optional.ofNullable(identifier);
+    }
+
+    public static DependentService constructDependentService(
```
`constructDependentService` -> `create(...)` for simplicity?

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,255 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Do we really need equality checks beyond the default object instance identity?

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,239 @@
+     * registered. It will not register the service if it was already registered.
+     */
+    @Override
+    public Object addingService(ServiceReference<Object> ref) {
```
Should we also handle service removal? I'd guess if a dependency disappears we should notify the service impl. sub-class. In the simple case I suppose it makes sense to stop the service. When the dependency re-appears the service will be restarted.

Services that need more complex dependency tracking logic will probably have to use custom logic (i.e. not extend `BaseActivator`)... WDYT?

tomekl007 replied:
From what I see, the use case that you described (restart service) is not implemented in any existing `Activator`.
I am ok with having that, but I think this should be a follow-up ticket. I would like this PR to have only refactoring changes without adding new features.

dimas-b replied:
sgtm

tomekl007 commented:
@dimas-b I've migrated the API to type-safe according to your suggestion. It looks pretty solid now. Please take a look.
Once we merge this PR, I will create a follow-up ticket to migrate all existing `Activator`s to this new API.
It should be done separately because all other activators have no unit or integration tests, so migrating them will require a substantial amount of manual testing. I've migrated only `HealthCheckerActivator` and `GraphqlActivator` because they were simple to test, and I wanted to have N use-cases for this new `BaseActivator` class in this PR.
Once we merge this PR, all new activators should use this new mechanism (I will adapt the CDC and Kafka CDC producer activators)

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,227 @@
+      return new ServicePointer<>(className, null);
+    }
+
+    public T getService() {
```
nit: maybe simply `get` since the class is a "service pointer" already?

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  public static class ServicePointer<T> {
+    private Class<T> className;
```
`className`?.. maybe `expectedClass`?

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -0,0 +1,227 @@
+  public synchronized void stop(BundleContext context) throws Exception {
+    if (started) {
+      logger.info("Stopping {}", activatorName);
+      stopService();
```
We automatically register the service, but do not automatically unregister it... this looks a bit awkward to me.

How about changing the sub-class lifecycle from `createService/stopService` to `startService/stopService` and delegate registration to sub-classes? This should also maturally suppose use cases where the sub-class registers more than one service object (from the same set of dependencies), which may soon be the case on the `feature-sql` branch.

tomekl007 replied:
The `Tracker` extends `ServiceTracker` and I think it handles the de-registration automatically:
```
	public void removedService(ServiceReference<S> reference, T service) {
		context.ungetService(reference);
	}
```

tomekl007 replied:
Re: registering in the child classes. I think that hiding this logic internally is quite convenient because you don't need to worry about it nor test it. For the N services use case, maybe it would be better to just rework the `BaseActivator` logic for this use case?

dimas-b replied:
sgtm.

tomekl007 replied:
I've added de-registration of target service: https://github.com/stargate/stargate/pull/370/commits/d39e9e92f95d2672bf78635957e1173536fa2767

dimas-b reviewed (commented):
LGTM overall :+1:  just one minor comment about create/stop vs. start/stop lifecycle (above).

dimas-b commented on core/src/main/java/io/stargate/core/activator/BaseActivator.java:
```diff
@@ -97,15 +100,30 @@ String constructDependenciesFilter() {
+    if (targetServiceRegistration != null) {
+      ServiceReference<?> reference = targetServiceRegistration.getReference();
+      logger.info("Unget service {} from {}", reference.getBundle(), activatorName);
+      context.ungetService(reference);
```
Should we also unget the dependencies?

tomekl007 replied:
No, they should be automatically `unget` by the `ServiceTracker`
